crates/bevy_hierarchy/src/child_builder.rs

```diff
@@ -481,21 +503,172 @@ impl<'w> BuildWorldChildren for EntityMut<'w> {
         });
         self
     }
+
+    fn set_parent(&mut self, parent: Entity) -> &mut Self {
+        let child = self.id();
+        self.world_scope(|world| {
+            world.entity_mut(parent).add_child(child);
+        });
+        self
+    }
+
+    fn remove_parent(&mut self) -> &mut Self {
+        let child = self.id();
+        if let Some(parent) = self.remove::<Parent>().map(|p| p.get()) {
+            self.world_scope(|world| {
+                remove_from_children(world, parent, child);
+                world.send_event(HierarchyEvent::ChildRemoved { child, parent });
+            });
+        }
+        self
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use super::{BuildChildren, BuildWorldChildren};
-    use crate::prelude::{Children, Parent};
+    use crate::{
+        components::{Children, Parent},
+        HierarchyEvent::{self, ChildAdded, ChildMoved, ChildRemoved},
+    };
     use smallvec::{smallvec, SmallVec};
 
     use bevy_ecs::{
         component::Component,
         entity::Entity,
+        event::Events,
         system::{CommandQueue, Commands},
         world::World,
     };
 
+    /// Assert the (non)existence and state of the child's [`Parent`] component.
+    fn assert_parent(world: &mut World, child: Entity, parent: Option<Entity>) {
+        assert_eq!(world.get::<Parent>(child).map(|p| p.get()), parent);
+    }
+
+    /// Assert the (non)existence and state of the parent's [`Children`] component.
+    fn assert_children(world: &mut World, parent: Entity, children: Option<&[Entity]>) {
+        assert_eq!(world.get::<Children>(parent).map(|c| &**c), children);
+    }
+
+    /// Used to omit a number of events that are not relevant to a particular test.
+    fn omit_events(world: &mut World, number: usize) {
+        let mut events_resource = world.resource_mut::<Events<HierarchyEvent>>();
+        let mut events: Vec<_> = events_resource.drain().collect();
+        events_resource.extend(events.drain(number..));
+    }
+
+    fn assert_events(world: &mut World, expected_events: &[HierarchyEvent]) {
+        let events: Vec<_> = world
+            .resource_mut::<Events<HierarchyEvent>>()
+            .drain()
+            .collect();
+        assert_eq!(events, expected_events);
+    }
+
+    #[test]
+    fn add_child() {
+        let world = &mut World::new();
+        world.insert_resource(Events::<HierarchyEvent>::default());
+
+        let [a, b, c, d] = std::array::from_fn(|_| world.spawn_empty().id());
+
+        world.entity_mut(a).add_child(b);
+
+        assert_parent(world, b, Some(a));
+        assert_children(world, a, Some(&[b]));
+        assert_events(
+            world,
+            &[ChildAdded {
+                child: b,
+                parent: a,
+            }],
+        );
+
+        world.entity_mut(a).add_child(c);
+
+        assert_children(world, a, Some(&[b, c]));
+        assert_parent(world, c, Some(a));
+        assert_events(
```

Comment:
What are these events for?

Storing the parent child relationship in a vec makes sense, but then the only logic that really makes sense to me is removing/adding to the vec. What do the events accomplish?

Reply:
oki doki, makes sense, thank :)